### PR TITLE
[OSPRH-32382] openstack_network_exporter calls trigger error messages in /var/log/messages and /var/log/openvswitch/ovs-vswitchd.log

### PR DESCRIPTION
**Pull request by abhiramnarayana**
## Summary

This PR fixes the 'Broken pipe' error messages that appear in /var/log/messages and /var/log/openvswitch/ovs-vswitchd.log when the openstack-network-exporter collects OpenFlow statistics from OVS. The root cause was that `getFlowStats()` only read a single OpenFlow message without handling multi-part replies, and both stats functions closed connections without draining pending messages (like ECHO requests), causing OVS to receive EPIPE when writing to the prematurely closed socket.

### OpenFlow Stats Collection (`openflow/openflow.go`)
- Added `drainPendingMessages()` helper function that reads and discards any pending OpenFlow messages from the connection before closing, using a 100ms timeout to prevent blocking indefinitely
- Modified `getFlowStats()` to loop reading multi-part OpenFlow replies until the `OFPSF_REPLY_MORE` flag is not set, accumulating flow stats from all reply parts
- Added `drainPendingMessages()` call to both `getFlowStats()` and `GetAggregateStats()` before connection close
- Extracted `readFlowStatsReplies()` internal function to enable unit testing of multi-part reply handling logic

### Unit Tests (`openflow/openflow_test.go`)
- Added comprehensive tests for `drainPendingMessages()` covering: timeout behavior, ECHO request draining, multiple messages, messages with body content, partial header handling, and read error handling
- Added `TestGetFlowStats_MultiPartReply` to verify multi-part reply handling with OFPSF_REPLY_MORE flag
- Added `TestGetFlowStats_SingleReply` as regression test for single-part replies
- Added `TestGetFlowStats_MultiPartReply_StopsAtFinal` to verify reading stops at final part and doesn't consume extra messages (like pending ECHO requests)
- Added `TestGetAggregateStats_DrainsPendingMessages` to verify drain pattern works correctly

## Implementation Notes

- The goloxi library (`github.com/skydive-project/goloxi/of10`) already provides the `OFPSFReplyMore` constant for checking multi-part reply flags, accessed via the `GetFlags()` method on `NiciraFlowStatsReply`
- The 100ms drain timeout is sufficient for Unix sockets without significantly impacting Prometheus scrape latency, while being long enough to catch most pending ECHO requests
- The `drainPendingMessages()` function intentionally ignores errors since timeout is expected when no messages are pending, and connection errors during drain are acceptable since we're about to close anyway
- Multi-part reply tests use empty `NiciraFlowStatsReply` messages due to a goloxi serialization quirk, but this doesn't affect production since OVS provides properly formatted responses

## Testing

- `go build ./...` - Full project builds successfully
- `go test -v ./openflow/...` - All 11 unit tests pass
- `gofmt -d openflow/openflow.go openflow/openflow_test.go` - No formatting issues
- `go vet ./openflow/...` - No issues
- `golangci-lint run ./openflow/...` - 0 issues

## Related Tickets

- [OSPRH-32382](https://redhat.atlassian.net/browse/OSPRH-32382)
- [OSPRH-33668](https://redhat.atlassian.net/browse/OSPRH-33668)

---
*Generated by [Forge](https://github.com/forge-sdlc/forge) SDLC Orchestrator*

### CODE REVIEW

**Review by aharivel — CHANGES_REQUESTED** [review]
## Review Summary

The fix itself is correct and addresses a real problem: EPIPE errors from OVS when the exporter closes OpenFlow connections without draining pending messages (ECHO requests). The multi-part reply handling is also a valid fix for large flow tables.

However, there are several issues that should be addressed before merge.

## Issues

### Correctness

**1. `drainPendingMessages`: no validation on `length < 8`**

If a malformed OpenFlow header has `length < 8`, the expression `length - 8` wraps around (uint16), causing allocation of up to ~65KB. Not a security risk on a local Unix socket, but still a bug:

```go
length := binary.BigEndian.Uint16(buf[2:4])
if length > 8 {
```

Should also handle `length < 8` explicitly (return/break — it's a malformed message).

**2. `readFlowStatsReplies`: unbounded loop**

If OVS keeps sending replies with `OFPSFReplyMore` set (shouldn't happen, but we're parsing external data), this loops forever. Should add a reasonable iteration cap (e.g., 10000) as a safety net.

### Style / Project conventions

**3. Copyright attribution is wrong in test file**

```go
// Copyright (c) 2024 Robin Jarry
```

Robin Jarry did not write this file. The agent copied the header from `openflow.go`. Should be the actual author's name, or omit the copyright line (keep just the SPDX identifier).

**4. Excessive comments**

The existing codebase style is minimal comments (see `openflow.go`, `collectors/netvf/collector_test.go`). This PR has block comments on every function and inline comments on obvious operations:

```go
// Discard the message and continue draining
// More parts follow, continue the loop
// This was the final part
// Accumulate stats from this reply part
// Set a short read deadline to avoid blocking indefinitely
// Attempt to read header
```

These should be removed to match project style.

### Dead code / test issues

**5. `mockConn.readTimeout` field is unused** — dead field, remove it.

**6. `timeoutError.Temporary()` — deprecated since Go 1.18.** `net.Error.Temporary()` is deprecated. Not a blocker but worth removing.

**7. `TestGetAggregateStats_DrainsPendingMessages` is misleading** — it doesn't test `GetAggregateStats` at all. It just calls `drainPendingMessages` directly, which is already covered by other tests. Either rename it or make it a real integration test.

**8. `TestGetFlowStats_EmptyReply` duplicates `TestGetFlowStats_SingleReply`** — identical logic: single reply, flags=0, assert 0 stats. Remove one.

## Positive

- Core fix is correct: multi-part reply handling + drain before close
- `readFlowStatsReplies` extraction is clean
- `OFPSFReplyMore` flag check is proper OpenFlow 1.0
- Drain timeout of 100ms is reasonable for Unix sockets
- Test coverage for `drainPendingMessages` edge cases is good (partial header, read error, multiple messages)

---
*Note: This PR was generated by [Forge](https://github.com/forge-sdlc/forge) SDLC Orchestrator — the issues above are typical of agent-generated code (wrong attribution, over-commenting, dead code, duplicate tests).*

**Comment by abhiramnarayana** [comment]
Forge is addressing PR review feedback now. This status update is informational.

**Comment by abhiramnarayana** [comment]
Forge is addressing PR review feedback now. This status update is informational.

**Comment by abhiramnarayana** [comment]
Forge is addressing PR review feedback now. This status update is informational.

**Comment by abhiramnarayana** [comment]
Forge is addressing PR review feedback now. This status update is informational.

**Comment by abhiramnarayana** [comment]
Forge is addressing PR review feedback now. This status update is informational.

**Comment by abhiramnarayana** [comment]
Forge is addressing PR review feedback now. This status update is informational.

**Comment by abhiramnarayana** [comment]
Forge is addressing PR review feedback now. This status update is informational.

**Comment by abhiramnarayana** [comment]
Forge is addressing PR review feedback now. This status update is informational.

**Comment by abhiramnarayana** [comment]
Forge is addressing PR review feedback now. This status update is informational.

**Comment by abhiramnarayana** [comment]
Forge is addressing PR review feedback now. This status update is informational.

**Comment by abhiramnarayana** [comment]
Forge is addressing PR review feedback now. This status update is informational.

**Comment by abhiramnarayana** [comment]
Forge is addressing PR review feedback now. This status update is informational.

File: openflow/openflow_test.go [hunk 1]
```diff
@@ -0,0 +1,290 @@
...
+	binary.BigEndian.PutUint32(header[4:8], xid)
+	return header
+}
+
+// TestDrainPendingMessages_Timeout verifies drainPendingMessages() respects timeout
+// when no messages are pending
+func TestDrainPendingMessages_Timeout(t *testing.T) {
+	conn := newMockConn()
+	// No data available - should timeout quickly
+
```
Comment (aharivel) [inline openflow/openflow_test.go]: unnecessary empty line

**Review by aharivel — CHANGES_REQUESTED** [review]
minor change

**Comment by abhiramnarayana** [comment]
Forge is addressing PR review feedback now. This status update is informational.

**Comment by abhiramnarayana** [comment]
Forge is addressing PR review feedback now. This status update is informational.

**Comment by abhiramnarayana** [comment]
Forge is addressing PR review feedback now. This status update is informational.

**Comment by abhiramnarayana** [comment]
Forge is addressing PR review feedback now. This status update is informational.

**Comment by abhiramnarayana** [comment]
Forge is addressing PR review feedback now. This status update is informational.

**Comment by abhiramnarayana** [comment]
Forge is addressing PR review feedback now. This status update is informational.

**Comment by abhiramnarayana** [comment]
Forge is addressing PR review feedback now. This status update is informational.

**Comment by abhiramnarayana** [comment]
Forge is addressing PR review feedback now. This status update is informational.

**Comment by abhiramnarayana** [comment]
Forge is addressing PR review feedback now. This status update is informational.

**Comment by abhiramnarayana** [comment]
Forge is addressing PR review feedback now. This status update is informational.

**Comment by abhiramnarayana** [comment]
Forge is addressing PR review feedback now. This status update is informational.

**Comment by abhiramnarayana** [comment]
Forge is addressing PR review feedback now. This status update is informational.

**Comment by abhiramnarayana** [comment]
Forge is addressing PR review feedback now. This status update is informational.

**Comment by abhiramnarayana** [comment]
Forge is addressing PR review feedback now. This status update is informational.

**Comment by abhiramnarayana** [comment]
Forge is addressing PR review feedback now. This status update is informational.

**Comment by abhiramnarayana** [comment]
Forge is addressing PR review feedback now. This status update is informational.

**Comment by abhiramnarayana** [comment]
Forge is addressing PR review feedback now. This status update is informational.

**Comment by abhiramnarayana** [comment]
Forge is addressing PR review feedback now. This status update is informational.

**Comment by abhiramnarayana** [comment]
Forge is addressing PR review feedback now. This status update is informational.

**Review by aharivel — CHANGES_REQUESTED** [review]
Remaining nits:
- drainPendingMessages still has its 4-line block comment (function-level). readFlowStatsReplies too. Rest of openflow.go has zero function-level comments. Minor but still doesn't match project style.
- Test file still over-commented compared to collector_test.go style (block comment on every test function).

Verdict: Correctness issues all fixed. Style is better but still verbose.

**Comment by aharivel** [comment]
also commit log with :
Signed-off-by: Forge <forge@example.com>
Doesn't mean anything. Please remove it.

**Review by aharivel — APPROVED** [review]
All correctness issues from the previous review have been addressed:

- `length < 8` validation added to `drainPendingMessages`
- `readFlowStatsReplies` loop bounded with `maxIterations`
- Wrong copyright attribution removed
- Dead code removed (`readTimeout` field, `Temporary()`, duplicate test)
- Misleading test removed

LGTM. Minor: test function comments are still verbose compared to project style, but not a blocker.
